### ISSUE TITLE
Default values of Kubelet's EvictionHard

### ISSUE THREAD

**Issue by vaibhav2107** [kubernetes/kubernetes#119985]
### What happened?

As we know, the default values of kubelet hard eviction's parameters are:
```
memory.available: "100Mi"
nodefs.available: "10%"
nodefs.inodesFree: "5%"
imagefs.available: "15%
```
When we change one of the default values, then the other unchanged parameters are set to 0. As per the code, this is the expected case.

### What did you expect to happen?

 The unchanged parameters should inherit their default values or is there any good scenario by setting the other parameters to 0.

### How can we reproduce it (as minimally and precisely as possible)?

NA

### Kubernetes version

<details>

```console
$ kubectl version
v1.25
```

</details>

### Cloud provider

<details>
NA
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by kannon92** [comment]
This behavior is expected.  

https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#hard-eviction-thresholds

In general it is recommend if you change one of those items you should change all the other ones.

**Comment by kannon92** [comment]
I realize I’m starting what you said but I don’t think this can be considered a bug.  I’d say it’s a feature to allow for different fields for HardEviction.

/remove-kind bug
/kind feature

**Comment by lowang-bh** [comment]
can we add a switch to controll it continue on inheriting the default values or be set to zero?

**Comment by vaibhav2107** [comment]
@kannon92 

> In general it is recommend if you change one of those items you should change all the other ones.

Yes, but if someone is changing only one parameter then all the other will be set to zero in that case.


> I realize I’m starting what you said but I don’t think this can be considered a bug. I’d say it’s a feature to allow for different fields for HardEviction.

Are you doing some work on this part?

**Comment by kannon92** [comment]
I meant stating not starting. I’m not working on this but I think this could be a welcome addition if someone wants to look into it.

the main issue is we set defaults if a user doesn’t specify those settings.

**Comment by kannon92** [comment]
Related issue: https://github.com/kubernetes/kubernetes/issues/121809

@vaibhav2107 you want to contribute a "fix" for this? Looking into this, I think we substitute user provided values in for the defaults.  

I think your suggestion could be a good addition.

**Comment by vaibhav2107** [comment]
@kannon92 I tthink I don't know exactly about the fix for this but I am investigating and trying to get it.
> @vaibhav2107 you want to contribute a "fix" for this? Looking into this, I think we substitute user provided values in for the defaults.

**Comment by vaibhav2107** [comment]
As per the comment and my understanding, the idea is to add switch or add a flag that can switch between the current behaviour and the behaviour after the PR change that I explained above.

Like if we add a flag inherit-eviction-default, 
If the flag is disabled and we change any parameter, then the default values are set as per the current behaviour.
	  

> inherit-eviction-default: false
> 	memory.available: "200Mi"

then the config will be set as
```
memory.available:  "200Mi" 
nodefs.available:  "0%" 
nodefs.inodesFree: "0%" 
imagefs.available: "0%"
imagefs.inodesFree: "0%"
```

If the flag is enabled and we change any parameter, then the default values are set as per the new behaviour

> 	inherit-eviction-default: true
> 	memory.available: "200Mi"

then the config will be set as
```
memory.available:  "200Mi" 
nodefs.available:  "10%" 
nodefs.inodesFree: "5%" 
imagefs.available: "15%" 
imagefs.inodesFree: "5%"
```

**Comment by vaibhav2107** [comment]
@SergeyKanzhelev @kannon92 
What could be the better approach to go forward with the issue. Why is the current behaviour considered as expected and do we need to keep the current behaviour as well, like what are the use cases for the current behaviour. I think if there are some cuse cases then we can go for the switch based approach else we can go for the simple change in beahviour similar to which the PR is open.

**Comment by haircommander** [comment]
we chatted about this at sig node. I believe this is from the kubelet's config reading code, where it can't tell the difference between explicitly 0 values and not present values, so it assumes the former. To fix this, we'd need to interpret 0 values as the default, which would mean we'd need to see whether 0 value is meaningful in the context. I have volunteered to look deeper into it.

**Comment by vaibhav2107** [comment]
Hi @haircommander, do you find time to look at this, how can we proceed further. Is zero value meaningful in the context?

**Comment by vaibhav2107** [comment]
@haircommander I found this PR regarding the 0% threshold value of Eviction hard parameters, as the PR description says that 0% threshold value doesn't make sense then why are we setting the default parameters to 0 in case of changing any of the parameter. Or is there any difference between the zero value that we are setting explicitly and the one that is set by default when changing the parameter. If there is none then I think we can go with the proposed change to inherit the default values to the unchanged parameters.

**Comment by atoato88** [comment]
@vaibhav2107 

> I found this PR regarding the 0% threshold value of Eviction hard parameters, as the PR description says that 0% threshold value doesn't make sense ...

Which PR do you refer?
Could you share the link for that?

**Comment by vaibhav2107** [comment]
@atoato88 oh, forgot to attach the link, I have updated my comment to add that

**Comment by haircommander** [comment]
hey I'm sorry I dropped the ball on this one. I think the key of that PR and why you are correct that we can take 0 as not present, is we also interpret 100% as "disable". I think we would need to update the documentation to say 100% is disable, and 0 will inherit the values of the base config (if all configs are 0, then we'll use 0 and disable). I do imagine there are some users that have disabled with 0 and thus would have the change in behavior, but I think this is better than disabling eviction when just one field is added...

**Comment by vaibhav2107** [comment]
> I do imagine there are some users that have disabled with 0 and thus would have the change in behavior, but I think this is better than disabling eviction when just one field is added...

To benefit these users who are using the current behaviour, we can use the switch case like I have explained in this [comment](https://github.com/kubernetes/kubernetes/issues/119985#issuecomment-2145561679), which can keep the both behaviours and could be helpful for the users.
How about it,,I would like to know the opinion on this.

**Comment by haircommander** [comment]
Yeah I think making it toggleable is the right approach. We may be able to get away with just a feature gate, with the result of the GA being the new behavior is the only option (giving admins ample warning and time to update their configuraitons).

**Comment by vaibhav2107** [comment]
Here is the summary of the issue to date. As the current behaviour is mentioned in the [bug description](https://github.com/kubernetes/kubernetes/issues/119985#issue-1853901529), I proposed the change in behaviour so that instead of zero value, eviction hard parameters should inherit their default values in case any of the parameter is changed. 
[Here](https://github.com/kubernetes/kubernetes/issues/119985#issuecomment-2145561679), I have explained how we can keep the both behaviours using the feature gate so that user does have the option to choose between the current and new behaviour. The community [has endorsed](https://github.com/kubernetes/kubernetes/issues/119985#issuecomment-2236628816) this idea and I am currently working on its implementation.
